File: day15.py
def solve(generator_a, generator_b, iter_count, part2):
    count = 0
    for i in range(iter_count):
        while True:
            generator_a = (generator_a * 16807) % 2147483647
            if not part2 or generator_a % 4 == 0:
                break
        while True:
            generator_b = (generator_b * 48271) % 2147483647
            if not part2 or generator_b % 8 == 0:
                break
        if (generator_a & 65535 == generator_b & 65535):
            count += 1
    return count


print(solve(289, 629, 40000000, False))
print(solve(289, 629, 5000000, True))

File: day15.go
package main

import "fmt"

func solveDay15(generator_a int, generator_b int, iter_count int, part2 bool) int {
	count := 0
	for i := 0; i <= iter_count; i++ {
		for true {
			generator_a = (generator_a * 16807) % 2147483647
			if !part2 || generator_a%4 == 0 {
				break
			}
		}

		for true {
			generator_b = (generator_b * 48271) % 2147483647
			if !part2 || generator_b%8 == 0 {
				break
			}
		}

		if generator_a&65535 == generator_b&65535 {
			count++
		}

	}
	return count
}

func main() {
	fmt.Println(solveDay15(289, 629, 40000000, false))
	fmt.Println(solveDay15(289, 629, 5000000, true))

}
